pmlc/dialect/tile/transforms/reorder_inputs.cc

```diff
@@ -0,0 +1,128 @@
+// Copyright 2020, Intel Corporation
+
+#include "mlir/IR/Builders.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Support/DebugStringHelper.h"
+#include "llvm/ADT/SetVector.h"
+
+#include "pmlc/dialect/layer/ir/ops.h"
+#include "pmlc/dialect/tile/ir/ops.h"
+#include "pmlc/dialect/tile/transforms/pass_detail.h"
+#include "pmlc/util/layout.h"
+#include "pmlc/util/logging.h"
+
+using namespace mlir; // NOLINT
+
+namespace pmlc::dialect::tile {
+
+namespace {
+
+int64_t getArgPos(Operation &op, BlockArgument &arg) {
+  auto argPos = 0;
+  for (size_t i = 0; i < op.getOperands().size(); i++) {
+    if (op.getOperand(i) == arg) {
+      argPos = i;
+    }
+  }
+  return argPos;
+}
+
+void performDataReordering(OpBuilder &builder, Operation *op, BlockArgument arg,
+                           MLFramework framework, bool isConst) {
+  // Check if op is a BoxOp, if so scan which argument is used,
+  // and get its tensor type.
+  // Later on create the new contraciton op with the same sizes and additional
+  // layout tag. These ops would be moved to init as weights reorders in the
+  // later pass.
+  auto layerOp = dyn_cast<layer::BoxOp>(op);
+  if (!layerOp)
+    return;
+
+  auto argPos = getArgPos(*op, arg);
+  auto tensorType =
+      op->getOperand(argPos).getType().dyn_cast<mlir::RankedTensorType>();
+  if (!tensorType)
+    return;
+
+  Type elementType = tensorType.getElementType();
+  auto ident = tile::createIdentity(builder, op->getLoc(), elementType,
+                                    AggregationKind::assign);
+  auto outRank = tensorType.getRank();
+  SmallVector<AffineExpr, 4> dimExprs;
+  dimExprs.reserve(outRank);
+
+  for (unsigned i = 0; i < outRank; i++) {
+    dimExprs.push_back(mlir::getAffineDimExpr(i, op->getContext()));
+  }
+  auto idMap = AffineMap::get(outRank, 0, dimExprs, op->getContext());
+
+  std::stringstream ss;
+  ss << "reordered_" << arg.getArgNumber();
+
+  auto newOp = builder.create<ContractionOp>(
```

Comment:
I don't quite understand this code.  Is it adding identity contractions between each arg of the box op and all of it's uses?  I imagine these would turn into transposes basically during lowering to pxa, which is good for the weights, but for internal activations I'd imagine we would want to add layout data directly so we don't induce transposes inline, or do those somehow optimize away.  Maybe some more lit tests would make a clear, I'm having a hard time reasoning about the whole pipeline herre.

Reply:
Yes, the idea was to create the contractions that would be moved to the init function on hoisting pass in tile dialect, and set their layout types. As you said those would be only transposes. This way we have these separated from the main function.  If it was not there then we will not read weights efficiently in the main function kernels, but have them transposed in main flow (not efficient). The reorderLayouts that is called after subgroup pass would modify the layout maps of those contractions instead of adding new ones. Yes the reordering of the activation is redundant, however I kept it to know that the pipeline would work. I was especially worried if any of the pass does not use strides and will have the analysis broken. With this explicit reorder on activation we kind of use old path with nxc for the analysis.